### DISCUSSION TITLE
Allow custom folders to be watched by Remix compiler.

### DISCUSSION THREAD

**Discussion by andresgutgon** [remix-run/remix#2074] [Proposals]
# What?
Hi, I'm doing a npm package and I wanted to have in my repo a Remix app to try my code. I wanted to do changes in my packages and view the changes in my remix example app.

The way I have setup my repo is like this. Is a [Rush](https://rushjs.io/) monorepo.

```
examples
  |-- basic (Remix App)
  |-- node_modules
     |-- @my-package
        |--core
           |--dist
              |--index.js
packages
  core
    |--dist
      |--index.js
    |--src
```
So when I do a change in `packages/core/src/index.ts` it's compiled in `packages/core/dist/index.js` ✅ 
Also if I inspect `examples/basic/node_modules/@my-packages/core/dist/index.js` I see the change ✅ 

But when I use the code in a Remix route like this:
``` javascript
import { hello } from "@my-package/core"
const loader = async () => {
  const sayHello = hello()
  return { hello: sayHello }
}
```
I don't see the function `hello` result updated. This is jus a dummy method
``` javascript
export const hello = (): string => "Hello there"
```

## 💡 So the idea
What If we could pass a custom folder in remix config to be watched. This would be super helpful for people developing `node` packages to be the changes reflected. I this it would be possible passing the option to Remix compiler.

I think here is where [watching is happening in Remix compiler](https://github.com/remix-run/remix/blob/4ad0234b3db09d6d77471c1e7a5b6e81714331ee/packages/remix-dev/compiler.ts#L213)

## MAYBE IS MORE SIMPLE
Please let me know if watching for changes in a Remix App `node_modules` folder is easier than this and I'm missing something.

**Comment by ccssmnn** [comment]
Checkout #1193 and especially [this comment](https://github.com/remix-run/remix/issues/1193#issuecomment-1033366665) for a working patch, which allows you to set custom paths for watching.

**Reply by andresgutgon** [reply]
Looks great! 💪 

Sadly I'm using `pnpm` and [patch-package](https://github.com/ds300/patch-package/pull/302) doesn't work with this package manager 😅 

But changing the files [commented here](https://github.com/remix-run/remix/issues/1193#issuecomment-1033366665) looks like it works. At least for making the Remix Watcher aware of the changes.

![image](https://user-images.githubusercontent.com/49499/155130975-fed371b4-e9dd-4ece-be34-981fa0f22fbf.png)


But that doesn't reflect on the code of my app. Looks it's using the old code

**Reply by andresgutgon** [reply]
I have to change something inside `app` folder of my remix code to view the changes in `node_modules/@mypackage` reflected in the UI

**Reply by kiliman** [reply]
`esbuild` caches the results of the files it compiles. So if none of the files in /app change, it won't see a need to compile. All the patch does is trigger the rebuild, but it doesn't affect the build itself.

The best solution, without patching Remix, is to have a file watcher watch for changes in node_modules and then update a file in /app to trigger the rebuild.

Luckily it's pretty simple.
```bash
npm install -D nodemon
```
Edit package.json. This runs nodemon and watches the folder. Any time a file in that folder changes, it will run the command to `touch app/refresh.ignore`. This is just a file that triggers the rebuild.
```json
"scripts": {
  "watch": "nodemon --watch node_modules/util --on-change-only --exec \"touch app/refresh.ignore\" --config nodemon-watch.json"
}
```
Add file nodemon-watch.json. *By default nodemon ignores default folders, including node_modules, so need to reset the ignore list.*
```json
{
  "ignoreRoot": [".git"]
}
```

**Reply by andresgutgon** [reply]
wow! looks super clean. I'll try, thanks @kiliman !

**Reply by kiliman** [reply]
BTW: nodemon only watches a single folder, so if you need multiple subfolders, you'll need to add extra --watch arguments. You could also put all this in the config file.

See nodemon docs https://github.com/remy/nodemon#nodemon

**Reply by andresgutgon** [reply]
It's not refreshing the content of the `node_modules`
I have running in 2 tabs in tmux: 1. Remix server and 2. Nodemon watcher
![image](https://user-images.githubusercontent.com/49499/155301941-e38e55de-4de7-4081-8af2-1ad651336a5f.png)

**Reply by andresgutgon** [reply]
Both are reacting to the change in my package.

And if I check inside generated `node_modules/@mypackage/core/dist/index.js` I see the new change
![image](https://user-images.githubusercontent.com/49499/155302202-bc3d51cb-e581-484d-bf9a-bf0b3ee7d20e.png)

This is the real code in my package:
![image](https://user-images.githubusercontent.com/49499/155302241-c8d2e940-b940-4ee0-9f78-8cc5bf9ce688.png)

**Reply by andresgutgon** [reply]
But in the web I see the last version when I started remix server.
![image](https://user-images.githubusercontent.com/49499/155302333-7bc5ec61-fba9-42d5-94a8-b2a962b406ab.png)

**Reply by andresgutgon** [reply]
So this is what I see:

1. When I change something in my package my build step in that package regenerates `dist` folder with new content
2. Remix app is notified of that change through nodemon watcher `touch app/refresh.ignore` and Remix server react to it
![image](https://user-images.githubusercontent.com/49499/155302636-14e8d877-84b6-4228-830e-fd6a418eadd0.png)


But some how the content of `node_modules/@mypackage/core/dist` is cached somewhere and new changes are not available until I kill remix server and run it again `pnpm run dev`

**Reply by andresgutgon** [reply]
Ok now is working if I put `nodemon` watcher in a pm2 file with a custom Express server:
``` javascript
module.exports = {
  apps: [
    {
      name: "Express App",
      script: "server.js",
      watch: ["remix.config.js", "app"],
      watch_options: {
        followSymlinks: false,
      },
      env: {
        NODE_ENV: "development",
      },
    },
    {
      name: "Remix",
      script: "remix watch",
      ignore_watch: ["."],
      env: {
        NODE_ENV: "development",
      },
    },
    {
      name: "Watch dependencies change",
      script: "nodemon -e js,d.ts --watch node_modules/@remix-storage --on-change-only --exec \"touch app/refresh.ignore\" --config nodemon-watch.json",
      watch: false,
      env: { NODE_ENV: "development" }
    },
  ],
}
```

**Reply by andresgutgon** [reply]
I don't understand what's the different of running `pnpm run dev` and `nodemon` or running both together through https://www.npmjs.com/package/pm2

But it works. It kills the server and restart it

**Reply by kiliman** [reply]
I see. The issue is that Remix App Server uses an internal Express server. When it rebuilds, it simply purges the require cache for the server bundle. However, by default the server bundle does not bundle the node_modules. This means the Express server still referenced the old one until restarted.

By using pm2, it will kill and restart the Express server anytime the watched files change. You have the "app" folder being watched. This is going to cause the server to restart every time you make an app code change, which may be inefficient. 

You may want to change the watch folder to just the node_module package you want. Then it will only restart when that file changes. That way you can let the Express server just purge the cache for the build stuff as usual.

**Reply by andresgutgon** [reply]
Is what I have, isn't it?
```
script: "nodemon -e js,d.ts --watch node_modules/@remix-storage --on-change-only --exec \"touch app/refresh.ignore\" --config nodemon-watch.json",
```
Maybe I'm not understanding

**Reply by kiliman** [reply]
No, I mean this:
```js
module.exports = {
  apps: [
    {
      name: "Express App",
      script: "server.js",
      watch: ["node_modules/@remix-storage"],
      watch_options: {
        followSymlinks: false,
      },
      env: {
        NODE_ENV: "development",
      },
    },
```
This way Express will only restart if the node_module changed instead of every app change.

**Reply by andresgutgon** [reply]
I tried this
![image](https://user-images.githubusercontent.com/49499/155513895-3330ca40-f30f-4548-b2c1-e05bbf5e0b6c.png)


Doesn't pick the changes in `node_modules`. I need `nodemon`

**Reply by kiliman** [reply]
Sorry if I wasn't clear. I didn't mean to remove the nodemon entry. I just meant to remove "app" from the Express watch list. That way it only restarts when you change the node_module. The Remix compiler will pickup the changes in app.

**Reply by andresgutgon** [reply]
Ahh! got it
